Spawn subimage jobs

annehaley opened the pull request.

annehaley committed:
Move subimage selection to job spawn pane

annehaley committed:
Include subimage datasets in investigation

annehaley committed:
Use django-large-image to crop subimage dataset

annehaley committed:
Fix cropping endpoint and add post_delete receiver

annehaley committed:
Make tiff cropping use all channels

annehaley committed:
Set subimage datasets to rootDataset

annehaley committed:
Allow user to specify and spawn a job

annehaley committed:
Jobs spawn successfully

annehaley committed:
Fix tests

annehaley committed:
Return jsonschema for job types directly from API

annehaley committed:
Modify test for job types endpoint

annehaley committed:
Specify job type input schema within module

annehaley committed:
Lint fix

waxlamp reviewed (commented):
I've seen this working and it's a good improvement. When @mcovalt approves via a formal code review, then this will be ok with me as well.

annehaley committed:
Lint fixes

mcovalt commented on web/src/components/JobPane.vue:
```diff
@@ -0,0 +1,219 @@
+}
+</style>
+
+<script>
```
In other components we use TypeScript. @waxlamp Is this still something we want to continue requiring in Phase I?

annehaley replied:
The conversion is simple enough, we may as well continue typescript. Changed in [d238ad3](https://github.com/atlascope/atlascope/pull/140/commits/d238ad3d22371cf03a2fe54af101f51cbeee748e)

waxlamp replied:
(Sticking with TypeScript is a good choice even for Phase I, because it requires a relatively small up-front investment of time to avoid relatively large time losses later.)

mcovalt commented on atlascope/core/models/dataset.py:
```diff
@@ -1,8 +1,15 @@
 from django.db import models
+from django.dispatch import receiver
 from django_extensions.db.models import TimeStampedModel
+from large_image_converter import _output_tiff
```
I'm a little nervous about importing a "private" (prefixed by underscore) method. We should definitely pin the version of `large_image_converter` if we do this. Otherwise, is there an official public method we can use?

annehaley replied:
this private function was all that I could find to do what we need. Some public functions invoke this but do much more additional stuff. The module as a whole has most of its functions as private with very narrow public use cases. I pinned the working version of `large_image_converter`, though. For now that will have to do. [24b2e61](https://github.com/atlascope/atlascope/pull/140/commits/24b2e610665820fb01468468b58239c2ebdaf749)

waxlamp replied:
@manthey, can you comment on this? Could it be possible to make `_output_tiff()` an official part of the API?

manthey replied:
It looks like you are really trying to do what `tifftools.commands.tiff_concat` does.  I haven't digested what you are doing besides concatenating a set of tiled tiff images.

annehaley replied:
I opened https://github.com/atlascope/atlascope/pull/166 to use tiff_concat instead. It works well!

mcovalt commented:
I noticed a new [package-lock.json](https://github.com/atlascope/atlascope/blob/279a0caf2c04348c2489d2a0ba30f1e148e5e4d3/web/package-lock.json). Are we moving away from yarn in favor of npm? I think we should choose one or the other instead of holding onto multiple lock files.

annehaley commented:
> I noticed a new [package-lock.json](https://github.com/atlascope/atlascope/blob/279a0caf2c04348c2489d2a0ba30f1e148e5e4d3/web/package-lock.json). Are we moving away from yarn in favor of npm? I think we should choose one or the other instead of holding onto multiple lock files.

This was not intended, thank you for catching that. I must have used an npm install command somewhere along the way

annehaley committed:
Use typescript in JobPane component

annehaley committed:
Delete package-lock.json

annehaley committed:
Frontend type/lint fixes

annehaley committed:
Allow subimages of subimages

annehaley committed:
Pin large-image-converter to known working version

annehaley committed:
Fix frontend typing again

mcovalt committed:
Update yarn.lock

mcovalt committed:
Change pyvips `shrinkMode`

The libvips library that comes with Ubuntu 20.04 doesn't support the
default of 'nearest'.

mcovalt reviewed (approved):
Awesome! Works well. This is a pretty massive PR and I did my best to understand the code changes.

I added a commit, 8f293fa969861917b691c9e71b9e82476c54be16, to change the default method pyvips used to shrink the image. It seems libvips on different architectures do not support the default 'nearest' method of shrinking an image. The method 'mode' should be more broadly supported and produce functionally equivalent results.

Take a look at that commit and maybe test things out once more before merging.

waxlamp commented:
> > I noticed a new [package-lock.json](https://github.com/atlascope/atlascope/blob/279a0caf2c04348c2489d2a0ba30f1e148e5e4d3/web/package-lock.json). Are we moving away from yarn in favor of npm? I think we should choose one or the other instead of holding onto multiple lock files.
> 
> This was not intended, thank you for catching that. I must have used an npm install command somewhere along the way

Friendly reminder not to do things like `git add *` and other similar command anti-patterns, like `git commit -a` without reviewing what was added (don't know if either of those happened in this case, but I figured I'd trot out another of my Git PSAs).
